Alerting: Remove message from PagerDuty summary field

AndrewBurian opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:
Removes the `- message` from the summary field.

Adds it instead in the custom details object under the key `message`

This greatly improves the experience in the PagerDuty UI. The summary field is cut short well below the actual 1k limit and is unreadable, where as the custom details are more readily visible in the incident view.

**Which issue(s) this PR fixes**:
<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #16930

**Special notes for your reviewer**:
Originally I had suggested adding blank lines, but I forgot the custom details field is an objet not free text, so this is actually even better for automation.

AndrewBurian committed:
Remove alert message from summary

closes #16930

AndrewBurian commented:
Sober second thought on this, I'm worried this might be too naive a fix.

`message` is not a very collision resistant name. There could easily be a series with the same name that shows up in the metrics, and this would clobber that.

Since JSON property order isn't guaranteed, it's also likely that the message shows up in between series data instead of before or after, which is not ideal to read.

Dumb fix A: Make the key something collision resistant and sortable in the hopes that PD internal implementation will sort them. Something like `_grafanaMessage`. Hideous but quick.

Complex fix B: Move the series that are currently included into a nested object, and leave the message as a top level property. 
```json
{
  "queries": {
    "seriesA": 500,
    "seriesB": 10000,
  },
  "message": "someMessage"
}
```
This is a way nicer fix, but at a way higher risk of breaking anyone who's using the alert object programatically from PD's API.

My instinct is my fear around downstream dependancies is overblown, and few things will be programatically consuming the custom details from PD. Since there's already Grafana in the loop, chances are you'd just use a webhook from Grafana directly and not scrape PD's API. For that reason I'd say go B. But it is a bigger change and will change the visuals in PD. Luckily, humans are more flexible than machines on the matter.

AndrewBurian commented:
This is how the nested object would be rendered in PagerDuty's interface for reference.

It also confirms that the ordering of the JSON object is not necessarily preserved. SeriesC and D got switched around.

![Screen Shot 2020-05-18 at 7 27 55 PM](https://user-images.githubusercontent.com/4976968/82277826-b495ea80-993d-11ea-83d2-7907cf913076.png)

AndrewBurian commented:
And on mobile this is rendered even nicer.
![pd-mobile](https://user-images.githubusercontent.com/4976968/82278293-bad89680-993e-11ea-905c-7215b05a18f7.png)

bergquist commented:
I like the end result of #B better 

> My instinct is my fear around downstream dependancies is overblown, and few things will be programatically consuming the custom details from PD. 

My worries are that people configured rulesets in Pagerduty and trigger actions based on series names. https://support.pagerduty.com/docs/rulesets#section-global-rulesets If they do #A would not be a problem but #B would since the series are nested one level deeper. Considering this is Alerting we have to avoid breaking.

Another option would be to introduce a settings for this but I think that would be very confusing for many users. This is a tricky one. I'm not sure :) But I would really like to move the message from the summary to custom details.

AndrewBurian commented:
Yeah it's a shame this didn't make it into v7 or it could have just been in the breaking changelog.

I don't really see a way forwards that doesn't break rulesets other than something like option A. I could set the key to "CustomMessage" and also do some collision detection on it just in case someone has a series of the same name.

The in-between is to briefly do both. Have the nested queries object, as well as have the series in the root object, and changelog that the root objects are going away. Then revisit in a few months and pull it. It's still a breaking change but with some migration room for anyone using rulesets. I do really like the structure of option B, so it might be worth it. It's more flexible for anything else that wants to be added to the custom details later too.

Having a setting is also possible, and would be less confusing if it was an opt-in to re-enable deprecated behaviour temporarily. Like `[ ]Include queries in custom details root. (deprecated) (i)` and an info popout explaining the change in behaviour. Then if anyone is caught off-guard and breaks their alerting, they can get it back for a bit. That's generally my preferred way to deal with deprecations, assume no one reads the change log, and give them a few versions to reenable it while they migrate.

bergquist commented:
> Then if anyone is caught off-guard and breaks their alerting,

In most cases, I agree with this but I think alerting is a bit different. I don't want people to have to experience an outage due to bad routing before they can enable it. Enabling #B with a config and switching behavior at the next major release could be an option. (even if its quite far until then)

AndrewBurian committed:
Gate the functionality behind a toggle.

The toggle is default off and keeps the original functionality.
Toggled on, it moves the message to custom details, and nests
the match results under another object.

AndrewBurian committed:
Doc updates

AndrewBurian commented:
@bergquist restructured so the change is hidden behind a toggle.

I'm not sure about the failing test. Seems to not like the spelling of IAM in `docs/sources/features/datasources/cloudwatch.md`, but I didn't touch that file, nor can I find the `words_to_ignore.txt` list anywhere to add an exception. The functional changes all work.

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -118,10 +118,16 @@ To set up PagerDuty, all you have to do is to provide an integration key.
+Message in details | Removes the Alert message from the PD summary field and puts it into custom details instead (2)
 
-**Note:** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an alert  ...
+**Note (1):** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an al ...
```
```suggestion
>**Note:** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an alert overrides the global Severity set on the notification channel if it's a valid level.
```

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -118,10 +118,16 @@ To set up PagerDuty, all you have to do is to provide an integration key.
-**Note:** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an alert  ...
+**Note (1):** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an al ...
+
+**Note (2):** Using Message In Details will change the structure of the `custom_details` field in the PagerDuty Event.
```
```suggestion
>Using Message In Details will change the structure of the `custom_details` field in the PagerDuty Event.
```

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -118,10 +118,16 @@ To set up PagerDuty, all you have to do is to provide an integration key.
+**Note (1):** The tags `Severity`, `Class`, `Group`, and `Component` have special meaning in the [Pagerduty Common Event Format - PD-CEF](https://support.pagerduty.com/docs/pd-cef). If an alert panel defines these tag keys, then they are transposed to the root of the event sent to Pagerduty. This means they will be available within the Pagerduty UI and Filtering tools. A Severity tag set on an al ...
+
+**Note (2):** Using Message In Details will change the structure of the `custom_details` field in the PagerDuty Event.
+This may break any custom event rules in your PagerDuty rules if you rely on the fields in `payload.custom_details`.
```
```suggestion
This might break custom event rules in your PagerDuty rules if you rely on the fields in `payload.custom_details`.
```

oddlittlebird commented on docs/sources/alerting/notifications.md:
```diff
@@ -118,10 +118,16 @@ To set up PagerDuty, all you have to do is to provide an integration key.
+**Note (2):** Using Message In Details will change the structure of the `custom_details` field in the PagerDuty Event.
+This may break any custom event rules in your PagerDuty rules if you rely on the fields in `payload.custom_details`.
+Move any existing rules using `custom_details.myMetric` to `custom_details.queries.myMetric`.
+This behaviour will become the default in a future version of Grafana.
```
```suggestion
This behavior will become the default in a future version of Grafana.
```

AndrewBurian committed:
Doc fixes and formatting

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

AndrewBurian committed:
Merge branch 'master' into pd-custom-details

AndrewBurian commented:
#25013 fixed the tests. This is good to review now ✅

aknuds1 commented:
@bergquist @marefr Can you have a look?

bergquist commented on pkg/services/alerting/notifiers/pagerduty.go:
```diff
@@ -88,8 +100,17 @@ func (pn *PagerdutyNotifier) buildEventPayload(evalContext *alerting.EvalContext
 	customData := simplejson.New()
-	for _, evt := range evalContext.EvalMatches {
-		customData.Set(evt.Metric, evt.Value)
+	if !pn.MessageInDetails {
```
Could you please flip these if statements? it's preferable to avoid using negation.

bergquist reviewed (commented):
Tested it and it works as expected. 

One comment about the if statements but after that's been fixed this is ready to be merged.

AndrewBurian committed:
remove negated checks

bergquist reviewed (approved):
LGTM

Thank you for contributing :)
